backend/main/chapters/c07_nested_loops.py

```diff
@@ -399,14 +399,62 @@ def solution(self, players: List[str]):
             """
              )]
 
-    class crack_password_exercise(ExerciseStep):
+    class player_vs_player_bonus(ExerciseStep):
         """
-Well done! If you'd like a bonus challenge, remove the repetition and only output each pair of players once, e.g:
+Well done! If you'd like a bonus challenge, print only those pairs that are in
+the same left-to-right order as they are in `players`, starting with pairs containing the leftmost person
+in `players` and moving right; e.g:
 
     Alice vs Bob
     Alice vs Charlie
     Bob vs Charlie
 
+        """
+
+        def solution(self, players: List[str]):
+            for i in range(len(players)):
+                for j in range(len(players)):
+                    if i < j:
+                        print(players[i], 'vs', players[j])
+
+        hints = """
+This is similar to the previous exercise, you'll need a for loop inside a for loop like before.
+This time, in your loops you need to use `range` and `len`, and use indexing to access the list entries.
+Look at the desired output: `Alice vs Bob`: `Alice` comes before `Bob` in the `players` list.
+We don't want to print `Bob vs Alice` because `Bob` comes AFTER `Alice` in `players`.
+The only pairs we want to print are those where the left player comes before the right player in the list.
+How can we express this relation in terms of the list indexes of the two for-loops?
+Once you figure out the relation, you can express it with an `if` statement.
+                """
+
+        tests = [
+            (["Alice", "Bob", "Charlie"], """\
+Alice vs Bob
+Alice vs Charlie
+Bob vs Charlie
+            """
+             ),
+            (["Frank", "Dave", "Emily"], """\
+Frank vs Dave
+Frank vs Emily
+Dave vs Emily
+            """
+             )]
+
+    final_text = """
+Excellent! The solution goes like this:
+ 
+    players = ["Alice", "Bob", "Charlie"]
+    for i in range(len(players)):
+        for j in range(len(players)):
+            if i < j:
+                print(players[i], 'vs', players[j])
+    """
+
+
+class Exercises(Page):
```

Comment:
Page names need to be unique and descriptive on their own, not just in context of the chapter.

In any case I think this particular bonus exercise is sufficiently different from the original that it doesn't need to come right after it. Leave everything on one page and move the bonus exercise to the end. Then add a link to the next page.

Reply:
OK... so the bonus comes after the cracking password and upside down triangle exercises, to finish the page.